src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs

```diff
@@ -0,0 +1,352 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+
+namespace Microsoft.AspNetCore.Razor.Language.Syntax
+{
+    internal static class MarkupElementRewriter
+    {
+        public static RazorSyntaxTree Rewrite(RazorSyntaxTree syntaxTree)
+        {
+            var rewriter = new Rewriter();
+            var rewrittenRoot = rewriter.Visit(syntaxTree.Root);
+
+            var newSyntaxTree = RazorSyntaxTree.Create(rewrittenRoot, syntaxTree.Source, syntaxTree.Diagnostics, syntaxTree.Options);
+            return newSyntaxTree;
+        }
+
+        public static RazorSyntaxTree RemoveMarkupElement(RazorSyntaxTree syntaxTree)
+        {
+            var remover = new Remover();
+            var rewrittenRoot = remover.Visit(syntaxTree.Root);
+
+            var newSyntaxTree = RazorSyntaxTree.Create(rewrittenRoot, syntaxTree.Source, syntaxTree.Diagnostics, syntaxTree.Options);
+            return newSyntaxTree;
+        }
+
+        private class Rewriter : SyntaxRewriter
```

Comment:
Is there precendent for any of this? Like is this based on something or did you write it fresh?

Reply:
https://github.com/aspnet/Blazor/blob/master/src/Microsoft.AspNetCore.Blazor.Razor.Extensions/ComponentDocumentRewritePass.cs FYI

Reply:
> Is there precendent for any of this? Like is this based on something or did you write it fresh?

This is mostly fresh. Some of the malformed tag handling logic is very similar to  https://github.com/aspnet/Razor/blob/master/src/Microsoft.AspNetCore.Razor.Language/Legacy/TagHelperParseTreeRewriter.cs#L730